File: DsaWithCpp/Tree/BinaryTree/SpiralTraversal.cpp
#include <iostream>
#include <stack>
#include "Node.cpp"
#include "TreeBuilder.cpp"
using namespace std;

static void spiralTraversal(Node *root)
{
    if (root == NULL)
        return;
    stack<Node *> stk1, stk2;
    stk1.push(root);
    while (!stk1.empty() || !stk2.empty())
    {
        while (!stk1.empty())
        {
            Node *currentNode = stk1.top();
            stk1.pop();
            cout << currentNode->data << " ";
            if (currentNode->left != NULL)
                stk2.push(currentNode->left);
            if (currentNode->right != NULL)
                stk2.push(currentNode->right);
        }
        cout << endl;
        while (!stk2.empty())
        {
            Node *currentNode = stk2.top();
            stk2.pop();
            cout << currentNode->data << " ";
            if (currentNode->right != NULL)
                stk1.push(currentNode->right);
            if (currentNode->left != NULL)
                stk1.push(currentNode->left);
        }
        cout << endl;
    }
}

// int main()
// {
//     int nodes[] = {1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1};
//     Node *root = buildTree(nodes);
//     spiralTraversal(root);
// }

File: Python/tree/binary_tree/spiral_traversal.py
from node import Node
from tree_builder import build_tree


def spiral_traversal(root: Node):
    if root is None:
        return
    stk1, stk2 = [], []
    stk1.append(root)
    while len(stk1) > 0 or len(stk2) > 0:
        while len(stk1) > 0:
            node = stk1.pop()
            print(node.data, end=" ")
            if node.left is not None:
                stk2.append(node.left)
            if node.right is not None:
                stk2.append(node.right)
        print()
        while len(stk2) > 0:
            node = stk2.pop()
            print(node.data, end=" ")
            if node.right is not None:
                stk1.append(node.right)
            if node.left is not None:
                stk1.append(node.left)
        print()


def main():
    nodes = [1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1]
    root = build_tree(nodes)
    spiral_traversal(root)


if __name__ == "__main__":
    main()
